src/witan/send/adroddiad/transcode.clj

```diff
@@ -0,0 +1,165 @@
+          (add-hint RenderingHints/KEY_DITHERING           RenderingHints/VALUE_DITHER_DISABLE)
+          (add-hint RenderingHints/KEY_RENDERING           RenderingHints/VALUE_RENDER_QUALITY)
+          (add-hint RenderingHints/KEY_STROKE_CONTROL      RenderingHints/VALUE_STROKE_PURE)
+          (add-hint RenderingHints/KEY_FRACTIONALMETRICS   RenderingHints/VALUE_FRACTIONALMETRICS_ON)
+          (add-hint RenderingHints/KEY_TEXT_ANTIALIASING   RenderingHints/VALUE_TEXT_ANTIALIAS_OFF))
+        (.setRenderingHints renderer hints)
+        renderer))))
+
+(defn children [element]
+  (let [c (.getChildNodes element)]
+    (if (zero? (.getLength c))
+      nil
+      (map #(.item c %) (range (.getLength c))))))
+
+(defn ->xml
+  ([element]
+   (->xml element true))
+  ([element children?]
+   (if (instance? org.w3c.dom.Text element)
+     (.getWholeText element)
+     (let [attrs (.getAttributes element)]
+       (merge
+        {:tag (keyword (.getLocalName element))
+         :ns  (.getNamespaceURI element)}
+        (when (and attrs (not (zero? (.getLength attrs))))
+          {:attrs (into {}
+                        (map (fn [i] (let [attr (.item attrs i)]
+                                       [(keyword (.getName attr))
+                                        (.getNodeValue attr)]))
+                             (range (.getLength attrs))))})
+        (when children?
+          (when-let [c (children element)]
+            {:content (mapv #(->xml % children?) c)})))))))
+
+(defn- document-dimensions [doc]
+  (-> doc ->xml :content first :attrs
+      (select-keys [:width :height])
+      (update :width parse-double)
+      (update :height parse-double)))
+
+(def transcoder-keys
+  {:width      SVGAbstractTranscoder/KEY_WIDTH
+   :height     SVGAbstractTranscoder/KEY_HEIGHT
+   :max-width  SVGAbstractTranscoder/KEY_MAX_WIDTH
+   :max-height SVGAbstractTranscoder/KEY_MAX_HEIGHT})
+
+(defn vl-map->bytearray [vl-chart-map]
+  (-> vl-chart-map
+      json/json-str
+      darkstar/vega-lite-spec->svg
+      svg-string->document))
+
+(defn svg-document->png
+  ([svg-document]
+   (svg-document->png svg-document {}))
+  ([svg-document {:keys [filename width scale]}]
+   (with-open [out-stream (cond
+                            filename
+                            (FileOutputStream. filename)
+                            :else
+                            (ByteArrayOutputStream.))]
+     (let [{doc-width :width :as dimensions} (document-dimensions svg-document)
+           in (TranscoderInput. svg-document)
+           out (TranscoderOutput. out-stream)
+           trans (high-quality-png-transcoder)]
+       (cond
+         scale
+         (.addTranscodingHint trans (:width transcoder-keys) (float (* scale doc-width)))
+         width
+         (.addTranscodingHint trans (:width transcoder-keys) (float width)))
+       (.transcode trans in out)
+       (when (not filename)
+         (.toByteArray out-stream))))))
```

Comment:
I feel like these arities should be the other way around as clojure is a 1 pass compiler, but I might be mis-remembering if this is working.

Reply:
Yeah I used this arity structure from some other Clojure code